backend/execution_cache/src/mutation.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"log"
+
+	"k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ArgoWorkflowTemplate string = "workflows.argoproj.io/template"
+	ExecutionKey         string = "pipelines.kubeflow.org/execution_cache_key"
+	AnnotationPath       string = "/metadata/annotations"
+)
+
+var (
+	podResource = metav1.GroupVersionResource{Version: "v1", Resource: "pods"}
+)
+
+// mutatePodIfCached will check whether the execution has already been run before from MLMD and apply the output into pod.metadata.output
+func mutatePodIfCached(req *v1beta1.AdmissionRequest) ([]patchOperation, error) {
+	// This handler should only get called on Pod objects as per the MutatingWebhookConfiguration in the YAML file.
+	// However, if (for whatever reason) this gets invoked on an object of a different kind, issue a log message but
+	// let the object request pass through otherwise.
+	if req.Resource != podResource {
+		log.Printf("Expect resource to be %q, but found %q", podResource, req.Resource)
+		return nil, nil
+	}
+
+	// Parse the Pod object.
+	raw := req.Object.Raw
+	pod := corev1.Pod{}
+	if _, _, err := universalDeserializer.Decode(raw, nil, &pod); err != nil {
+		return nil, fmt.Errorf("could not deserialize pod object: %v", err)
+	}
+
+	var patches []patchOperation
+	annotations := pod.ObjectMeta.Annotations
+	template, exists := annotations[ArgoWorkflowTemplate]
+	var executionHashKey string
+	if !exists {
+		return patches, nil
+	}
+
+	// Generate the executionHashKey based on pod.metadata.annotations.workflows.argoproj.io/template
+	hash := sha256.New()
+	hash.Write([]byte(template))
```

Comment:
What happens if the template is nil?

Reply:
It will be handled in https://github.com/kubeflow/pipelines/pull/3036/files#diff-af9bdf231e9c75a4b00bb2b348aacf5bR59. If !exists

Reply:
Lack of entry is different form an entry with null value.

Reply:
I guess, this is impossible with annotations in Go. Unlike other languages, in Go a `map[string]string` object cannot have nil values.